README.md
# HAP

[![build status](https://rdc.hand-china.com/gitlab/HAP/hap/badges/master/build.svg)](https://rdc.hand-china.com/gitlab/HAP/hap/commits/master)

[3.x 开发文档](http://eco.hand-china.com/doc/hap/latest/)

[4.x 开发文档](https://wiki.choerodon.com.cn/bin/view/O-Choerodon/P-HAP%E6%A1%86%E6%9E%B6%E7%BB%84/HAP4.0%E5%BC%80%E5%8F%91%E6%89%8B%E5%86%8C/)


## 项目模块

- hap-front-kendo kendoUI 相关部分
- hap-front-html 原kendoUI 实现的功能界面
- hap 框架汇总模块 (包含其他分模块 提供 HAP 完整功能)
- hap-core 框架核心模块 (包含组织架构、系统管理等基础功能)
- hap-gateway 服务管理 (在接口管理基础上 重新优化设计 推荐使用服务管理)
- hap-interface 接口管理 (原是从其他系统移植过来的功能)
- hap-job 计划任务 (基于Quartz管理 自动执行)
- hap-mail 邮件服务 (基于JavaMail)
- hap-oauth2 Oauth认证服务 (基于OAuth2)
- hap-report 报表服务 (基于UReport2)
- hap-task 任务管理 (基于HAP权限控制 手动执行)
- hap-workflow 工作流 (基于Activiti)
- hap-attachment 附件管理
- hap-cxf-ws 发布WebService (基于CXF)
- hap-security-standard 标准登录 (基于spring-boot-starter-security)
- hap-security-cas CAS单点登录 (基于spring-security-cas)
- hap-security-ldap LDAP认证 (基于spring-security-ldap)
- hap-modeling 建模平台(已弃用)

## 创建数据库

默认数据库为MySQL，修改MySQL的配置文件 my.cnf (or my.ini),在 [mysqld] 下添加
lower_case_table_names=1
character_set_server=utf8
max_connections=500

执行以下SQL，创建默认数据库
CREATE SCHEMA hap_dev DEFAULT CHARACTER SET utf8;
CREATE USER hap_dev@'%' IDENTIFIED BY 'hap_dev';
GRANT ALL PRIVILEGES ON hap_dev.* TO hap_dev@'%';
FLUSH PRIVILEGES;


## 初始化数据库

根目录下执行如下命令（ Linux/Mac 可以直接执行，Windows可安装Git后使用Git Bash执行）
```sh
    sh init-database.sh
```

### 脚本参数介绍
- Dspring.datasource.url： 数据源url
- Dspring.datasource.username：数据库登陆用户
- Ddata.drop=false：是否清除数据库数据
- Ddata.init：是否使用excel数据进行数据初始化
- Ddata.dir：groovy和excel所在的文件夹，扫描指定文件夹下的所有.groovy和.xlsx文件
- Ddata.mode : 数据初始化的模式可选：normal（默认，初始化groovy和excel），iam（初始化权限数据，仅单体模式需要），all（二者并集）
- Ddata.update.exclusion：初始化数据库更新数据的时候忽略的表或列，忽略表直接写表名，表名.列名 表示忽略某一列，使用逗号分割

## 启动前准备
   修改配置文件application.yml 文件配置数据库与相关连接信息。示例如下：
```yml
spring:
  datasource:
    password: ${MYSQL_PASS:hap_dev}
    url: jdbc:mysql://${MYSQL_HOST:localhost}:${MYSQL_PORT:3306}/${MYSQL_DB:hap_dev}?useSSL=false&allowPublicKeyRetrieval=true
    username: ${MYSQL_USER:hap_dev}
redis:
  ip: ${REDIS_IP:localhost}
  db: ${REDIS_DB:10}
db:
  type: mysql
mybatis:
  identity: JDBC
```

  如果没有依赖安全模块（hap-security-standard、hap-security-cas、hap-security-ldap）
   
  IRequest中默认配置了admin用户的信息，可在application.yml中自定义配置，示例如下

```yml
default:
  security:
    userId: 10002
    userName: jessen
    roleId: 10001
    allRoleId: 10002,10001
    companyId: 10001
    employeeCode: JESSEN
    locale: en_GB
```

  权限校验时，如果permission未注册，默认会拦截并返回无权限

  可在application.yml中配置是否跳过权限校验，示例如下(线上环境无需配置 默认强制进行权限校验,本地开发可设置为true 方便调试)

```yml
default:
  permission:
    access: true 
```
## 本地开发

如果是本地开发，建议单独启动前后端

### 单独启动后端项目
hap目录启动所有模块，各个模块下单独启动对应服务，命令如下：
```mvn
    mvn spring-boot:run
```
可以使用postman等工具 访问localhost:8080进行测试

### 单独启动前端项目
hap目录下执行,命令如下：
```text
    mvn clean package
    npm install --registry https://nexus.choerodon.com.cn/repository/choerodon-npm/
    npm start
```
注：如果是linux系统，在npm start失败时可以先尝试给node_modules增加权限
```
chmod -R 755 node_modules
```
目前如果只启动前端是无法登陆访问界面的，
需要在config.js里面配置proxyTarget（本地或者远程后端服务器地址），
如proxyTarget: 'http://localhost:8080'

然后浏览器访问 localhost:9090 访问HAP界面

## 部署项目

如果是要测试下项目整体运行或者部署项目，建议前后端一起运行

### 前后端一起运行
- 首先在 hap 目录运行 npm run build 打包前端
- 在 hap 目录执行以下  mvn 命令：
```mvn
    mvn spring-boot:run
```
浏览器访问 localhost:8080 访问HAP界面


CHANGELOG
2016-08-31
----------
1. 修改 mysql 版 sys_profile.sql 建表脚本 : 主键被错误设置为 VARCHAR
2. 添加 mybatis-configuration.xml 配置文件, 设置 jdbcTypeForNull 为 NULL
   解决 某些 Oracle 数据库 插入/更新 null 值报错的问题


2016-08-29
----------
1. 启用 objectVersionNumber 字段
   查询时,会返回  objectVersionNumber 字段
   对通用 Mapper,Service 的影响:
       插入时,会自动设置 objectVersionNumber 为 1,并返回
       更新时,where 条件 追加 OBJECT_VERSION_NUMBER=#{objectVersionNumber} (仅当 objectVersionNumber 不为 null 时有效)
          当 update 返回 0,并且 objectVersionNumber 不为 null 时,将抛出异常 UpdateFailedException
       删除时,变动同上

   当发生 UpdateFailedException 时, 前台会看到错误消息: 记录不存在或版本不一致。记录 ID:{0}
   此消息需要手动在 "描述维护" 中注册:error.record_not_exists_or_version_not_match

   对于非通用 Service, 不受影响, 但也可自己判断 update count, 抛出异常


2016-08-25
----------
1. 优化功能定义循环的错误
   当功能定义存在循环时,功能分配页面将显示错误消息,消息中包含简要的循环链。可据此消息排查错误。
2. oauth2 与 CAS 同时使用时, oauth2 可以使用用户名密码获取 token (之前无法使用 secret 模式)

2016-08-23
----------
1. 升级框架通用查询 sql 的排序功能
    现在根据 BaseDTO 中 sortname,sortorder 来动态生成排序 sql；
    同时会结合主键，来保证排序的唯一性；
    自动生成的 sql 会对 sortname,sortorder 进行校验，阻止 sql 注入；

2016-08-22
----------
1. 去除 附件上传 的 contentType 验证 (该验证 会导致有些类型无法上传)

2016-08-18
----------
1. 添加 checkboxlist 事件触发: onChangeValue

2016-08-16
----------
1. 修改 ProfileMapper.xml , ResultMap 继承错误
2. 升级 js,css : 修正一个低版本 IE 的兼容性问题

2016-08-11
----------
1. 删除 noeditor 样式 (会影响 grid 行的背景色)
2. 点击 grid 单元格不在触发 update 事件(之前会导致翻页时有提醒)
3. 添加 grid 单元格 编辑状态角标, checkBox render 也会支持

2016-08-10
----------
1. 添加 favcicon.png
2. 调整登录、角色选择页面字体样式

2016-08-08
----------
1. 删除角色时, 现在会删除相关的资源分配(角色分配,功能分配)

2016-08-02
----------
1. 增加 freemarker 自定义 bean 的支持 : 注解 @FreeMarkerBean
    将需要在 freemarker 中使用的类 加上此注解, 并通过 spring 定义为 bean 即可。
    引用的 name 即为 bean 的 name。(手动注册 bean 的话, 需要指定 name)

2016-08-01
----------
1. 修复一个 bug：不使用 standard security 时报错。
2. 静态资源文件现在可以被浏览器缓存。
    如果在调试时不希望被缓存，可以利用浏览器的开发者工具，禁用浏览器缓存。

2016-07-31
----------
1. 修正一个 IE 下的 BUG：当部署 hap 在根节点时，无法获取验证码的 cookie。
2. 精简一些不必要的代码

2016-07-29
----------
1. 增加 checkbox 事件：onChange，参考 demo.html
2. 现在 prompt 中含有（单、双）引号会被自动转换成 html 转义字符
    这个改变不会影响数据库中的值，仅仅在使用之前进行转换。
    这在之前，会导致 json 语法错误。
3. 对于数据多语言，sys_multilanguage_editor.html 有相应的修改


2016-07-28
----------
1. 修复一个 bug：功能资源分配会分页，并删除没有查询出来的数据。
    现在不再分页，并且增量更新变化的资源。


2016-07-27
----------
1. 调整一些样式，主要包括：登录，角色选择，主页
2. 统一添加 X-UA-Compatible  IE=edge,chrome=1
3. 增加了 checkbox(render)，checkboxlist 的 demo

2016-07-22
----------
1. 优化权限校验：
   a) 已经 permitAll 的 url 不再走自定义校验
   b) 现在可以注册（并校验）带有 占位符 的 url ，例如 file/{fileId}/view


2016-07-20
----------
1. 修复一个 bug：功能缓存只会影响当前语言环境。这个 bug 可能会导致一些菜单显示问题，以及新加功能的权限的问题。

2016-07-18
----------
1. 优化日期解析,现在处理空字符串(日期)时不在报错
2. 修复 sys_code_edit.html 多语言编辑器空白的 bug

2016-07-15
----------
1. DTO _token 校验，现在根据 主键 是否有值，来决定是否跳过校验（之前是根据__status）


2016-07-10
----------
1. 调整 Grid 行 line-height:28px ,这将会使 grid 行中的内容上下居中（之前是偏上）。
2. 增强用户管理页面，增加 状态、有效日期、修改密码 等功能
3. hap.js 增加一个 render 函数,用于渲染 grid 中的 checkbox 编辑器
   Hap.createGridCheckBoxRender(),具体用法可以参考 sys_resource.html
   LigerUI 的 checkbox 编辑器在 grid 中问题很多,暂时没法用。
   此次修改的 基于 render 的伪编辑器,已知存在一个问题:无法使用 tab 键获得焦点。

2016-07-07
----------
1. 增加登录异常支持
   使用 login-faliure-handler 取代 login-faliure-url
   用户被锁定(LOCKED):抛出异常 error.user.account_locked
   用户尚未激活(未到有效时间):抛出异常 error.user.account_not_active
   用户已过期(已经超出有效时间):抛出异常 error.user.account_expired
   以上三个异常对应的消息 目前尚未包含在标准版本中,需手动注册。
   
2. 修复验证码 url 截断问题(有时验证码报404)

2016-07-05
----------
1. 调整首页右上角菜单,用户菜单中的3个子菜单经常点不中的问题
   调整菜单项 line-height:30px;display:inline-block
   
2016-07-04
----------
1. 修改 head.html,使用浏览器时区作为默认时区(如果系统中没有设置)
2. 修改 role.html,当只有一个角色时,自动选择并登录
3. 添加 FontAwesome 标准图标支持, 除了 icon-xx,现在还可以使用 fa-xx
4. 修复一个拼写错误,这个错误导致保存功能分配之后,缓存被清空,访问所有页面404

2016-06-29
----------
1. 修复一个 html 路径自动映射问题:当 url 有4段时,无法正确取到 html 文件路径
2. hap.js 添加 两个用于日期 和 日期+时间的 render
   Hap.gridDateRender    : yyyy-MM-dd
   Hap.gridDateTimeRender: yyyy-MM-dd HH:mm:ss

2016-06-28
----------
1. 添加登录成功事件监听机制,接口:IAuthenticationSuccessListener
   实现此接口,并在 spring context 中注册的 bean,将在标准登录成功以后被通知

2016-06-27
----------
1. 修复退出后的登录页面为空白的问题

2016-06-26
----------
1. 添加注解:ExtensionAttribute
   将此注解加在 DTO 类上,表示此 DTO 对应的表不需要 标准的16(1+15)个扩展字段
   
2016-06-23
----------
1. 修复 grid 中 comboBox 点中后显示内容为空的问题
   移除 grid 列中配置的 textField
2. 增强 prompt 维护页面可用性
   可以连续点击多次新增按钮,语言环境字段将自动填充
   为新增的行填写 prompt_code 时,将自动复制到相关联的行。


hap-core/react/src/sys/role/view/RoleMsg.js
import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { inject, observer } from 'mobx-react';
import querystring from 'query-string';
import remove from 'lodash/remove';
import _ from 'lodash';
import { Icon, Button, Form, Input, Table, Tooltip, Tabs, Checkbox } from 'choerodon-ui';
import { FormattedMessage, injectIntl } from 'react-intl';
import { axiosPro as axios, ContentPro as Content, Header, Page, closeTab, getCurrentTab } from '@choerodon/boot';
import { RESOURCES_LEVEL } from '@choerodon/boot/lib/containers/common/constants';
import { get, set } from 'mobx';
import '../Role.scss';
import Sider from './Sider';
import RoleStore from '../stores/RoleStore';

const { TabPane } = Tabs;
const FormItem = Form.Item;
const { TextArea } = Input;
const intlPrefix = 'global.role';
const LEVEL_NAME = {
  site: '全局层',
  organization: '组织层',
  project: '项目层',
  user: '个人中心',
};

@Form.create({})
@withRouter
@injectIntl
@inject('AppState')
@observer
export default class CreateRole extends Component {
  constructor(props) {
    super(props);
    const queryObj = querystring.parse(props.location.search);
    this.level = queryObj.level || undefined;
    this.base = queryObj.base ? queryObj.base.split(',') : [];
    this.roleId = queryObj.roleId || undefined;
    this.isEdit = !!this.roleId;
    // eslint-disable-next-line prefer-destructuring
    this.tabLevel = RESOURCES_LEVEL.split(',')[0];

    props.cacheLifecycles.didRecover(this.componentDidRecover.bind(this));
  }

  componentDidRecover() {
    const queryObj = querystring.parse(this.props.history.location.search);
    if (!queryObj.level && !queryObj.base && !queryObj.roleId) {
      return;
    }
    this.level = queryObj.level || undefined;
    this.base = queryObj.base ? queryObj.base.split(',') : [];
    this.roleId = queryObj.roleId || undefined;
    this.isEdit = !!this.roleId;
    // eslint-disable-next-line prefer-destructuring
    this.tabLevel = RESOURCES_LEVEL.split(',')[0];
    if (!this.isEdit) {
      this.props.form.resetFields();
      RoleStore.setSelectedPermissions([]);
    }
    
    this.loadLabelsAndMenus();
  }

  componentDidMount() {
    if (!this.isEdit) {
      RoleStore.setSelectedPermissions([]);
    }
    this.loadLabelsAndMenus();
  }

  loadLabelsAndMenus = () => {
    const { tabLevel, base } = this;
    RoleStore.setTabLevel(tabLevel);
    this.loadMenu(RoleStore.tabLevel || tabLevel);
    if (base.length) {
      RoleStore.getSelectedRolePermissions(base)
        .then((res) => {
          RoleStore.setSelectedPermissions(res.map(p => p.id));
        });
    }
    if (this.isEdit) {
      RoleStore.getRoleById(this.roleId)
        .then((res) => {
          RoleStore.setRoleMsg(res);
          RoleStore.setSelectedPermissions(res.permissions.map(p => p.id));
        });
    }
  }

  loadMenu = (tabLevel) => {
    RoleStore.loadMenu(tabLevel)
      .then((menus) => {
        set(RoleStore.menus, tabLevel, menus.subMenus);
        set(RoleStore.expandedRowKeys, tabLevel, this.getAllIdByLevel(tabLevel));
        if (!RoleStore.tabLevel) {
          RoleStore.setTabLevel(tabLevel);
        }
      });
  }

  check = (selectedPermissions, menu, sign, type) => {
    if (menu.subMenus) {
      menu.subMenus.map(menuItem => this.check(selectedPermissions, menuItem, sign, type));
    }
    this.checkOne(selectedPermissions, menu, sign, type);
  }

  checkOne = (selectedPermissions, menu, sign, type) => {
    if (type === 'all') {
      if (menu.permissions.map(p => p.id).some(pid => selectedPermissions.findIndex(v => v === pid) === -1)) {
        sign.sign = false;
      }
    } else if (type === 'none') {
      if (menu.permissions.map(p => p.id).some(pid => selectedPermissions.findIndex(v => v === pid) !== -1)) {
        sign.sign = false;
      }
    }
  }

  getTabCodes = () => {
    const LEVEL_OBJ = {
      site: ['site', 'user'],
      project: ['project'],
      organization: ['organization'],
    };
    return LEVEL_OBJ[this.level] || [];
  }

  getIds = (menu, res) => {
    res.push(menu.id);
    if (menu.subMenus) {
      menu.subMenus.map(menuItem => this.getIds(menuItem, res));
    }
  }

  getAllIdByLevel = (level) => {
    const menus = get(RoleStore.menus, level) || [];
    const res = [];
    menus.map(menu => this.getIds(menu, res));
    return res;
  }

  getOneMenuPermissons = (menu, res) => {
    res.res = res.res.concat(menu.permissions.map(p => p.id));
  }

  getPermissions = (menu, res) => {
    if (menu.subMenus) {
      menu.subMenus.map(menuItem => this.getPermissions(menuItem, res));
    }
    this.getOneMenuPermissons(menu, res);
  }

  getAllPermissionsByRecord = (record, originRes) => {
    const res = originRes || { res: [] };
    this.getPermissions(record, res);
    if (!originRes) {
      res.res = [...new Set(res.res)];
      return res.res;
    }
  }

  getAllPermissionsByLevel = (level) => {
    const menus = get(RoleStore.menus, level) || [];
    const res = { res: [] };
    menus.map(menu => this.getAllPermissionsByRecord(menu, res));
    res.res = [...new Set(res.res)];
    return res.res;
  }

  getCheckState = (type, selectedPermissions, record) => {
    const sign = { sign: true };
    this.check(selectedPermissions, record, sign, type);
    return sign.sign;
  }

  checkCode = (rule, value, callback) => {
    const { isEdit } = this;
    if (isEdit) {
      callback();
      return;
    }
    const params = { code: value };
    axios.post('/v1/roles/check', JSON.stringify(params)).then((mes) => {
      if (mes.success === false) {
        const { intl } = this.props;
        callback(intl.formatMessage({ id: `${intlPrefix}.code.exist.msg` }));
      } else {
        callback();
      }
    });
  };

  linkToChange = (url, refresh) => {
    const { history } = this.props;
    history.push({
      pathname: url,
      state: {
        refresh: true,
      },
    });
  };

  handleExpand = (expanded, record) => {
    const expandedRowKeys = get(RoleStore.expandedRowKeys, RoleStore.tabLevel) || [];
    if (expanded) {
      expandedRowKeys.push(record.id);
    } else {
      remove(expandedRowKeys, v => v === record.id);
    }
    set(RoleStore.expandedRowKeys, RoleStore.tabLevel, expandedRowKeys);
  }

  handleCheckboxAllClick = (checkedAll, checkedNone, checkedSome, e) => {
    const allPermissionsByRecord = this.getAllPermissionsByLevel(RoleStore.tabLevel);
    const { selectedPermissions } = RoleStore;
    let sp = selectedPermissions.slice();
    if (checkedNone || checkedSome) {
      sp = sp.concat(allPermissionsByRecord);
      sp = [...new Set(sp)];
    } else {
      remove(sp, p => allPermissionsByRecord.includes(p));
    }
    RoleStore.setSelectedPermissions(sp);
  }

  handleCheckboxClick = (record, checkedAll, checkedNone, checkedSome, e) => {
    const allPermissionsByRecord = this.getAllPermissionsByRecord(record);
    const { selectedPermissions } = RoleStore;
    let sp = selectedPermissions.slice();
    if (checkedNone || checkedSome) {
      sp = sp.concat(allPermissionsByRecord);
      sp = [...new Set(sp)];
    } else {
      remove(sp, p => allPermissionsByRecord.includes(p));
    }
    RoleStore.setSelectedPermissions(sp);
  }

  handleCreate = (e) => {
    e.preventDefault();
    this.props.form.validateFieldsAndScroll((err) => {
      if (!err) {
        const role = {
          name: this.props.form.getFieldValue('name').trim(),
          code: this.props.form.getFieldValue('code').trim(),
          description: this.props.form.getFieldValue('description').trim(),
          level: this.level,
          permissions: RoleStore.selectedPermissions.slice().map(p => ({ id: p })),
          objectVersionNumber: RoleStore.roleMsg.objectVersionNumber,
        };
        const { intl } = this.props;
        if (!role.permissions || !role.permissions.length) {
          Choerodon.prompt('该角色没有一个权限，请至少选择一个权限');
          return;
        }
        if (this.isEdit) {
          RoleStore.editRoleByid(this.roleId, role)
            .then((data) => {
              if (!data.failed && data.id) {
                Choerodon.prompt(intl.formatMessage({ id: 'modify.success' }));
                this.linkToChange('/hap-core/role');
              } else {
                Choerodon.prompt(data.message || '保存失败');
              }
            })
            .catch((error) => {
              Choerodon.prompt(error || '保存失败');
            });
        } else {
          RoleStore.createRole(role)
            .then((data) => {
              if (data && data.id) {
                Choerodon.prompt(intl.formatMessage({ id: 'create.success' }));
                this.linkToChange('/hap-core/role', true);
                closeTab();
              } else {
                Choerodon.prompt(data.message);
              }
            })
            .catch((errors) => {
              if (errors.response.data.message === 'error.role.roleNameExist') {
                Choerodon.prompt(intl.formatMessage({ id: `${intlPrefix}.name.exist.msg` }));
              } else {
                Choerodon.prompt(intl.formatMessage({ id: 'create.error' }));
              }
            });
        }
      }
    });
  };

  handleReset = () => {
    this.linkToChange('/hap-core/role');
    closeTab();
  };

  handleChangeTabLevel = (key) => {
    RoleStore.setTabLevel(key);
    if (!get(RoleStore.menus, key)) {
      this.loadMenu(key);
    }
  }

  handleOpenSider = (record) => {
    RoleStore.setCurrentMenu(record);
    RoleStore.setSiderVisible(true);
  }

  handleSiderOk = (selectedPermissions) => {
    const { isEdit } = this;
    const isDefault = isEdit && RoleStore.roleMsg.builtIn;
    if (isDefault) {
      RoleStore.setSiderVisible(false);
      return;
    }
    RoleStore.setSelectedPermissions(selectedPermissions);
    RoleStore.setSiderVisible(false);
  }

  handleSiderCancel = () => {
    RoleStore.setSiderVisible(false);
  }

  handleClickExpandBtn = () => {
    const tabLevel = RoleStore.tabLevel || this.tabLevel;
    const expand = get(RoleStore.expand, tabLevel);
    if (expand) {
      // 需要展开
      set(RoleStore.expandedRowKeys, tabLevel, this.getAllIdByLevel(tabLevel));
    } else {
      // 需要收起
      set(RoleStore.expandedRowKeys, tabLevel, []);
    }
    set(RoleStore.expand, tabLevel, !expand);
  }

  renderCheckbox = (isDefault) => {
    const { selectedPermissions } = RoleStore;
    const allPermissionsByLevel = this.getAllPermissionsByLevel(RoleStore.tabLevel);
    const checkedAll = allPermissionsByLevel.every(p => selectedPermissions.includes(p));
    const checkedNone = allPermissionsByLevel.every(p => !selectedPermissions.includes(p));
    const checkedSome = !checkedAll && !checkedNone;
    return (
      <Checkbox
        indeterminate={checkedSome}
        onChange={this.handleCheckboxAllClick.bind(this, checkedAll, checkedNone, checkedSome)}
        checked={!checkedNone}
        disabled={isDefault}
      />
    );
  }

  renderForm = () => {
    const { props: { intl, form: { getFieldDecorator } }, isEdit } = this;
    const isDefault = isEdit && RoleStore.roleMsg.builtIn;
    const formItemLayout = {
      labelCol: {
        xs: { span: 24 },
        sm: { span: 100 },
      },
      wrapperCol: {
        xs: { span: 24 },
        sm: { span: 10 },
      },
    };

    return (
      <Form layout="vertical">
        <FormItem {...formItemLayout} style={{ display: 'inline-block', marginRight: 12 }}>
          {getFieldDecorator('code', {
            rules: [{
              required: true,
              whitespace: true,
              message: intl.formatMessage({ id: `${intlPrefix}.code.require.msg` }),
            }, {
              validator: this.checkCode,
            }],
            validateFirst: true,
            initialValue: isEdit ? RoleStore.roleMsg.code : undefined,
            validateTrigger: 'onBlur',
          })(
            <Input
              autoComplete="off"
              label={<FormattedMessage id={`${intlPrefix}.code`} />}
              size="default"
              style={{ width: 250 }}
              disabled={isEdit}
              maxLength={64}
              showLengthInfo={false}
            />,
          )}
        </FormItem>
        <FormItem {...formItemLayout} style={{ display: 'inline-block' }}>
          {getFieldDecorator('name', {
            rules: [{
              required: true,
              whitespace: true,
              message: intl.formatMessage({ id: `${intlPrefix}.name.require.msg` }),
            }],
            initialValue: isEdit ? RoleStore.roleMsg.name : undefined,
          })(
            <Input
              autoComplete="off"
              label={<FormattedMessage id={`${intlPrefix}.name`} />}
              style={{ width: 250 }}
              maxLength={64}
              showLengthInfo={false}
              disabled={isDefault}
            />,
          )}
        </FormItem>
        <FormItem {...formItemLayout} style={{ width: 512 }}>
          {getFieldDecorator('description', {
            initialValue: isEdit ? RoleStore.roleMsg.description : undefined,
          })(
            <TextArea
              autoComplete="off"
              label="描述"
              style={{ width: 512 }}
            />,
          )}
        </FormItem>
      </Form>
    );
  }

  renderTable = (level) => {
    const data = get(RoleStore.menus, level);
    const expandedRowKeys = get(RoleStore.expandedRowKeys, level) || [];
    const { isEdit } = this;
    const isDefault = isEdit && RoleStore.roleMsg.builtIn;
    const columns = [{
      title: '菜单',
      dataIndex: 'name',
      key: 'name',
      width: '35%',
      render: (text, record) => (
        <span>
          <Icon type={record.icon} style={{ marginRight: 8, verticalAlign: 'top' }} />
          {text}
        </span>
      ),
    }, {
      title: this.renderCheckbox(isDefault),
      dataIndex: 'id',
      key: 'id',
      width: '36px',
      render: (text, record) => {
        const { selectedPermissions } = RoleStore;
        const checkedAll = this.getCheckState('all', selectedPermissions, record);
        const checkedNone = this.getCheckState('none', selectedPermissions, record);
        const checkedSome = !checkedAll && !checkedNone;
        return (
          <Checkbox
            indeterminate={checkedSome}
            onChange={this.handleCheckboxClick.bind(this, record, checkedAll, checkedNone, checkedSome)}
            checked={!checkedNone}
            disabled={isDefault}
          />
        );
      },
    }, {
      title: '页面入口',
      dataIndex: 'route',
      key: 'route',
    }, {
      title: '',
      width: '50px',
      key: 'action',
      align: 'right',
      render: (text, record) => {
        const { subMenus } = record;
        if (!subMenus || !subMenus.length) {
          return (
            <Tooltip
              title={<FormattedMessage id="detail" />}
              placement="bottom"
            >
              <Button
                shape="circle"
                icon="predefine"
                size="small"
                onClick={this.handleOpenSider.bind(this, record)}
              />
            </Tooltip>
          );
        }
        return null;
      },
    }];
    return (
      <Table
        loading={false}
        filterBar={false}
        pagination={false}
        columns={columns}
        defaultExpandAllRows
        dataSource={data ? data.slice() : []}
        childrenColumnName="subMenus"
        rowKey={record => record.id}
        expandedRowKeys={expandedRowKeys.slice()}
        onExpand={this.handleExpand}
        indentSize={25}
      />
    );
  }

  renderBtns = () => {
    const { isEdit } = this;
    const isDefault = isEdit && RoleStore.roleMsg.builtIn;
    return (
      <div style={{ marginTop: 32 }}>
        {
          isDefault ? null : (
            <Button
              funcType="raised"
              type="primary"
              onClick={this.handleCreate}
              style={{ marginRight: 12 }}
            >
              <FormattedMessage id={!this.isEdit ? 'create' : 'save'} />
            </Button>
          )
        }
        <Button
          funcType="raised"
          onClick={this.handleReset}
          style={{ color: '#3F51B5' }}
        >
          <FormattedMessage id="cancel" />
        </Button>
      </div>
    );
  }

  renderSider = () => {
    const { siderVisible, currentMenu, selectedPermissions } = RoleStore;
    const { isEdit } = this;
    const isDefault = isEdit && RoleStore.roleMsg.builtIn;
    if (siderVisible) {
      return (
        <Sider
          selectedPermissions={selectedPermissions}
          menu={currentMenu}
          onOk={this.handleSiderOk}
          onCancel={this.handleSiderCancel}
          disabled={isDefault}
        />
      );
    }
    return null;
  }

  renderTab = () => {
    const tabLevel = RoleStore.tabLevel || this.tabLevel;
    const expand = get(RoleStore.expand, tabLevel);
    return (
      <React.Fragment>
        <div>
          <span style={{ marginRight: 80, fontSize: '16px' }}>菜单分配</span>
          <Button
            type="primary"
            funcType="flat"
            icon={expand ? 'expand_more' : 'expand_less'}
            onClick={this.handleClickExpandBtn}
          >
            全部{expand ? '展开' : '收起'}
          </Button>
        </div>
        <Tabs onChange={this.handleChangeTabLevel} activeKey={tabLevel}>
          {this.getTabCodes().map(level => (
            <TabPane tab={LEVEL_NAME[level] || level} key={level}>
              {this.renderTable(level)}
            </TabPane>
          ))}
        </Tabs>
      </React.Fragment>
    );
  }

  render() {
    return (
      <Page className="c7n-roleMsg">
        <Header
          title={`${!this.isEdit ? '创建' : '修改'}角色`}
          backPath="/hap-core/role"
        />
        <Content>
          {this.renderForm()}
          {this.renderTab()}
          {this.renderBtns()}
          {this.renderSider()}
        </Content>
      </Page>
    );
  }
}
